mlir/lib/Dialect/WebAssembly/IR/WebAssemblyOps.cpp

```diff
@@ -89,6 +89,56 @@ Block *BlockReturnOp::getTarget() {
       ->getSuccessor(0);
 }
 
+//===----------------------------------------------------------------------===//
+// ExtendFromLowBytesSOp
+//===----------------------------------------------------------------------===//
+
+ParseResult ExtendFromLowBytesSOp::parse(::mlir::OpAsmParser &parser,
+                                         ::mlir::OperationState &result) {
+  OpAsmParser::UnresolvedOperand operand;
+  uint64_t nBits;
+  auto parseRes = parser.parseInteger(nBits);
+  parseRes = parser.parseKeyword("low");
+  parseRes = parser.parseKeyword("bits");
+  parseRes = parser.parseKeyword("from");
+  parseRes = parser.parseOperand(operand);
+  parseRes = parser.parseColon();
+  Type inType;
+  parseRes = parser.parseType(inType);
+  if (!inType.isInteger())
+    return failure();
+  llvm::SmallVector<Value, 1> opVal;
+  parseRes = parser.resolveOperand(operand, inType, opVal);
+  if (parseRes.failed())
+    return failure();
+  result.addOperands(opVal);
+  result.addAttribute(
+      ExtendFromLowBytesSOp::getBitsToTakeAttrName(OperationName{
+          ExtendFromLowBytesSOp::getOperationName(), parser.getContext()}),
+      parser.getBuilder().getI64IntegerAttr(nBits));
+  result.addTypes(inType);
+  return success();
+}
+
+void ExtendFromLowBytesSOp::print(OpAsmPrinter &p) {
+  p << " " << getBitsToTake().getUInt() << " low bits from ";
+  p.printOperand(getInput());
+  p << ": " << getInput().getType();
+}
+
+LogicalResult ExtendFromLowBytesSOp::verify() {
+  auto bitsToTake = getBitsToTake().getValue().getLimitedValue();
+  if (bitsToTake != 32 && bitsToTake != 16 && bitsToTake != 8)
```

Comment:
Also if the input is i32 extend can't take i32 (I would be a no-op but the operation doesn't exist in Wasm)

Reply:
That's the next check iiuc your comment.